### ISSUE TITLE
Prediction dictionaries matching item to prediction

### ISSUE THREAD

**Issue by PabloAMC** [PyTorchLightning/pytorch-lightning#11064]
## 🚀 Feature

A way to match data and trainer prediction in 
```
predictions = trainer.predict(model, datamodule=datamodule)
```

### Motivation

Right now, even if one can use `predict` to predict the results from the datamodule it becomes impossible to know which prediction corresponds to each item.

### Pitch

Dictionary mapping identifier to prediction

### Alternatives

I have tried assuming that `datamodule.predict_data_frame` (in a tabular task) does not shuffle the data, but I think it somehow does. I would be great to have it fixed though.

### Additional context

The documentation says the output of predict is a list of dicts, but it's a list of lists.

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @borda @rohitgr7

**Comment by ananthsub** [comment]
Did you take a look at the Prediction writer callback? This could help unnest your batch to get indices corresponding to each sample: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/callbacks/prediction_writer.py

> Right now, even if one can use predict to predict the results from the datamodule it becomes impossible to know which prediction corresponds to each item.

However, in Lightning, what is returned by your dataloader is ultimately a black box to the framework. You could also consider including metadata in your batch to distinguish per-sample level information. Then your output predictions could include the input metadata so you can associate them later.

**Comment by PabloAMC** [comment]
Yes, that was @ethanwharris guess, to associate the metadata.
